### ISSUE TITLE
build: Pin the c++ toolchain versions for libprotobuf (bazel)

### ISSUE THREAD

**Issue by jlinder** [cockroachdb/cockroach#56054]
Do so while keeping the ability to pick up a C/C++ compiler for that specific version from outside of the bazel build tree, for example the C/C++ toolchain installed by an OS' package distribution. See conversation [here](https://cockroachlabs.slack.com/archives/C019A8NANH0/p1603731441391500).

**Comment by alan-mas** [comment]
Working on it :)

**Comment by alan-mas** [comment]
I am reading and checking the Bazel official documentation for [toolchain](https://docs.bazel.build/versions/master/toolchains.html) and I think I kind of know where to start. Also I did check the conversation between Otan and Raphael but I do have some doubts, hope you might help me:

I can see that libprotobuf is there present in `c-deps/BUILD.bazel`:
```
cmake_external(
    name = "libprotobuf",
    binaries = ["protoc"],
    cache_entries = {
        "CMAKE_BUILD_TYPE": "Release",
        "CMAKE_TARGET_MESSAGES": "OFF",
        "protobuf_BUILD_TESTS": "OFF",
    },
    lib_source = "@protobuf//:all",
    static_libraries = ["libprotobuf.a"],
    visibility = ["//visibility:public"],
    working_directory = "cmake",
)
```
Otan told me about that we need to make sure we build using the same [gcc and g++](https://stackoverflow.com/questions/172587/what-is-the-difference-between-g-and-gcc) (compiler-drivers of the GNU Compiler Collection)

Is there something I might be missing here?

**Comment by alan-mas** [comment]
Documenting all the info I have so far so everyone can be in synch on what is going on around here.

### What compiler is in the builder `./build/builder.sh`?
**TLDR:** 
- We are forcing Docker to run using clang (defining `update-alternatives` and `apt-get install clang`), but at the end of the day clang still needs gcc libraries to compile (that is why we can see g++ and gcc installed in the docker when the docker container is running)
- Bazel is using Clang (The one I do have at my local host)

**Complete story:**
From `builder` prespective:
if you run `build/builder.sh make libprotobuf` it will construct a `defs.mk` using your local values and will volume some paths (workdir, tempdir, etc) and it is going to run a container with that command you put before (`make libprotobuf`).
I got a little deeper to found that I was able to check which compiler docker container was using (changing `builder.sh` a bit) but did not find anything,  so I am going to assume (based on the Dockerfile steps for [removing gcc/g++](https://github.com/cockroachdb/cockroach/blob/master/build/builder/Dockerfile#L116) and then in [line 125 install clang](https://github.com/cockroachdb/cockroach/blob/master/build/builder/Dockerfile#L124-L125) and passing [update-alternatives](https://github.com/cockroachdb/cockroach/blob/master/build/builder/Dockerfile#L130) for clang) that it is using clang.

Bazel perspective:
If I run `bazel build //c-deps:libprotobuf --sandbox_debug`  and check inside `bazel-bin/c-deps/libprotobuf/logs/CMake.log`  I can see this:
```
-- The C compiler identification is AppleClang 10.0.1.10010046
-- The CXX compiler identification is AppleClang 10.0.1.10010046
```
Means that Bazel is using clang (AppleClang) also in my local I can see `Apple LLVM version 10.0.1 (clang-1001.0.46.4)` although I do have `gcc` and `g++` as well due to [clang by default uses gcc standard libraries in linux](https://github.com/compiler-explorer/compiler-explorer/issues/903) (Docker platform native).

### What's the status on pin a specific compiler?
**TLDR:** 
- Found 2 ways of explain Bazel to use a specific compiler tool, but the behavior seems pretty tricky. and already got 
- A third approach is already on track. It is to create a [custom tool chain rule](https://docs.bazel.build/versions/master/toolchains.html)

**Complete story:**
I’ve found some useful (and tricky information) during my tests while trying to pin toolchain for `g++`,  `gcc` and `clang`.
Seems that Bazel already integrated a flag for [specify one compiler](https://stackoverflow.com/questions/47624689/how-to-set-c-compiler-on-os-x-with-bazel) `BAZEL_USE_CPP_ONLY_TOOLCHAIN=1`
So I did some tests using different flag values at my local, where I need to add that I have these compilers inside `/usr/bin` :
```
llvm-g++ -> clang++
llvm-gcc -> clang
g++
gcc
clang
clang++
```
Commands I tried:
**1 - CC using g++**
```
BAZEL_USE_CPP_ONLY_TOOLCHAIN=1 bazel build --action_env CC=/usr/bin/g++ //c-deps:libprotobuf --sandbox_debug
```
“The CMAKE_C_COMPILER is set to a C++ compiler”
ERROR: https://stackoverrun.com/es/q/1970397

**2 - CC using gcc**
```
BAZEL_USE_CPP_ONLY_TOOLCHAIN=1 bazel build --action_env CC=/usr/bin/gcc //c-deps:libprotobuf --sandbox_debug
```
It is working but it is still using Applelang.
Log inside `bazel-bin/c-deps/libprotobuf/logs/CMake.log`:
```
-- The C compiler identification is AppleClang 10.0.1.10010046
-- The CXX compiler identification is AppleClang 10.0.1.10010046
```

**3 - CC using clang++**
```
BAZEL_USE_CPP_ONLY_TOOLCHAIN=1 bazel build --action_env CC=/usr/bin/clang++ //c-deps:libprotobuf --sandbox_debug
```
“The CMAKE_C_COMPILER is set to a C++ compiler”
Seems to be the same issue as the 1 run.

**4 - CC using clang**
```
BAZEL_USE_CPP_ONLY_TOOLCHAIN=1 bazel build --action_env CC=/usr/bin/clang //c-deps:libprotobuf --sandbox_debug
```
It is working but it is still using Applelang.
Log inside `bazel-bin/c-deps/libprotobuf/logs/CMake.log`:
```
-- The C compiler identification is AppleClang 10.0.1.10010046
-- The CXX compiler identification is AppleClang 10.0.1.10010046
```

**5 - CXX using g++**
```
BAZEL_USE_CPP_ONLY_TOOLCHAIN=1 bazel build --action_env CXX=g++ //c-deps:libprotobuf --sandbox_debug
```
It is working but it is still using Applelang
Log inside `bazel-bin/c-deps/libprotobuf/logs/CMake.log`
```
-- The C compiler identification is AppleClang 10.0.1.10010046
-- The CXX compiler identification is AppleClang 10.0.1.10010046
```

**6 - CXX using clang++**
```
BAZEL_USE_CPP_ONLY_TOOLCHAIN=1 bazel build --action_env CXX=clang++ //c-deps:libprotobuf --sandbox_debug
```
It is working but it is still using Applelang
Log inside `bazel-bin/c-deps/libprotobuf/logs/CMake.log`
```
-- The C compiler identification is AppleClang 10.0.1.10010046
-- The CXX compiler identification is AppleClang 10.0.1.10010046
```

**7 - CXX using clang**
```
BAZEL_USE_CPP_ONLY_TOOLCHAIN=1 bazel build --action_env CXX=clang //c-deps:libprotobuf --sandbox_debug
```
It is working but it is still using Applelang
Log inside `bazel-bin/c-deps/libprotobuf/logs/CMake.log`
```
-- The C compiler identification is AppleClang 10.0.1.10010046
-- The CXX compiler identification is AppleClang 10.0.1.10010046
```

**8 - CXX using gcc**
```
BAZEL_USE_CPP_ONLY_TOOLCHAIN=1 bazel build --action_env CXX=gcc //c-deps:libprotobuf --sandbox_debug
```
It is working but it is still using Applelang
Log inside `bazel-bin/c-deps/libprotobuf/logs/CMake.log`
```
-- The C compiler identification is AppleClang 10.0.1.10010046
-- The CXX compiler identification is AppleClang 10.0.1.10010046
```
In conclusion when we use `CXX` instead of `CC` as an input flag for bazel build [cmake is smart enough](https://github.com/snap-stanford/snap/issues/165) to understand that needs to compile C++ in the correct way.
With that approach I am able to "kind of" specify de compiler we do want Bazel (and Cmake) to use.


Another approach I tested was to change a bit Bazel build file `c-deps/BUILD.bazel` addings flags like:
```
        "CMAKE_C_COMPILER": "gcc",
        "CMAKE_CXX_COMPILER": "g++",
        "CMAKE_C_COMPILER_ID": "Clang",
        "CMAKE_CXX_COMPILER_ID": "Clang",
```

Both ID's flags are working but seems to be only like a "dummy name" and the other compilers are failing no matter which compiler I do use.
First they fail due to `protobuf_BUILD_TESTS` need to set to `OFF` even that inside the arguments you already set that to OFF in `c-deps/BUILD.bazel` [line 38](https://github.com/cockroachdb/cockroach/blob/master/c-deps/BUILD.bazel#L38), so if I go and changed directly in `c-deps/protobuf/cmake/CMakeLists.txt` - [Line 28](https://github.com/cockroachdb/protobuf/blob/fbafa9842933bc547fbe0dbca9c882f37ff3a526/cmake/CMakeLists.txt#L28)
Now passed that step but fails at another step:
```
CMake Error at cmake_install.cmake:41 (file):
  file INSTALL cannot copy file
  "/var/folders/zz/zyxvpxvq6csfxvn_n0000000000000/T/tmp.pVkHwnT2/libprotobuf-lite.a"
  to "/usr/local/lib/libprotobuf-lite.a": Operation not permitted.
```
This seems due to as Oliver commented:

> The error is atm we're using `make install`, and `make install` is trying to copy to `/usr/local/lib` instead of the local sandbox. We've worked around this by changing the install script, e.g. https://github.com/cockroachdb/cockroach/blob/master/c-deps/BUILD.bazel#L22
So by default it runs make && make install


###Next steps:
Do some investigation and working around a [custome toolchain creation(https://docs.bazel.build/versions/master/toolchains.html) as this seems to be the best idea, since we guarantee everyone's using the same CC version (sounds like otherwise we're depending on the user's OS).

##Useful links:
- https://github.com/bazelbuild/bazel/issues/4231
- https://cmake.org/cmake/help/latest/manual/cmake-compile-features.7.html
- https://gitlab.kitware.com/cmake/community/-/wikis/doc/cmake/Useful-Variables
- https://stackoverrun.com/es/q/1970397
- https://github.com/bazelbuild/bazel/blob/master/tools/cpp/cc_configure.bzl
- https://stackoverflow.com/questions/61292460/how-do-i-make-sure-that-my-default-c-c-compiler-is-gcc
- https://stackoverflow.com/questions/41356173/how-to-use-clang-instead-g-in-bazel
- https://github.com/bazelbuild/bazel/issues/4077
- https://github.com/bazelbuild/rules_foreign_cc
- https://github.com/otan-cockroach/rules_foreign_cc
- https://github.com/bazelbuild/rules_foreign_cc/blob/master/tools/build_defs/cmake.bzl
- https://github.com/dropbox/dbx_build_tools
- https://github.com/bazelbuild/rules_foreign_cc/blob/master/tools/build_defs/cmake.bzl#L83

**Comment by rickystewart** [comment]
#62739
